src/dataqueue/queue.h

```diff
@@ -0,0 +1,293 @@
+      uint64_t length);
+
+  static std::unique_ptr<Entry> CreateDataQueueEntry(
+      std::shared_ptr<DataQueue> data_queue);
+
+  static std::unique_ptr<Entry> CreateFdEntry(
+      BaseObjectPtr<fs::FileHandle> handle);
+
+  // Creates a Reader for the given queue. If the queue is idempotent,
+  // any number of readers can be created, all of which are guaranteed
+  // to provide the same data. Otherwise, only a single reader is
+  // permitted.
+  virtual std::shared_ptr<Reader> getReader() = 0;
+
+  // Append a single new entry to the queue. Appending is only allowed
+  // when isIdempotent() is false. v8::Nothing<bool>() will be returned
+  // if isIdempotent() is true. v8::Just(false) will be returned if the
+  // data queue is not idempotent but the entry otherwise cannot be added.
+  virtual v8::Maybe<bool> append(std::unique_ptr<Entry> entry) = 0;
+
+  // Caps the size of this DataQueue preventing additional entries to
+  // be added if those cause the size to extend beyond the specified
+  // limit.
+  //
+  // If limit is zero, or is less than the known current size of the
+  // data queue, the limit is set to the current known size, meaning
+  // that no additional entries can be added at all.
+  //
+  // If the size of the data queue is not known, the limit will be
+  // ignored and no additional entries will be allowed at all.
+  //
+  // If isIdempotent is true capping is unnecessary because the data
+  // queue cannot be appended to. In that case, cap() is a non-op.
+  //
+  // If the data queue has already been capped, cap can be called
+  // again with a smaller size.
+  virtual void cap(uint64_t limit = 0) = 0;
+
+  // Returns a new DataQueue that is a view over this queues data
+  // from the start offset to the ending offset. If the end offset
+  // is omitted, the slice extends to the end of the data.
+  //
+  // The slice will coverage a range from start up to, but excluding, end.
+  //
+  // Creating a slice is only possible if isIdempotent() returns
+  // true. This is because consuming either the original DataQueue or
+  // the new queue would change the state of the other in non-
+  // deterministic ways. When isIdempotent() returns false, slice()
+  // must return a nulled unique_ptr.
+  //
+  // Creating a slice is also only possible if the size of the
+  // DataQueue is known. If size() returns v8::Nothing<uint64_t>, slice()
+  // must return a null unique_ptr.
+  virtual std::shared_ptr<DataQueue> slice(
+      uint64_t start, v8::Maybe<uint64_t> end = v8::Nothing<uint64_t>()) = 0;
+
+  // The size of DataQueue is the total size of all of its member entries.
+  // If any of the entries is not able to specify a size, the DataQueue
+  // will also be incapable of doing so, in which case size() must return
+  // v8::Nothing<uint64_t>.
+  virtual v8::Maybe<uint64_t> size() const = 0;
+
+  // A DataQueue is idempotent only if all of its member entries are
+  // idempotent.
+  virtual bool isIdempotent() const = 0;
```

Comment:
https://github.com/nodejs/node/blob/main/doc/contributing/cpp-style-guide.md#camelcase-for-methods-functions-and-classes

Reply:
heh, too much switching between multiple projects with different naming conventions. sigh.